k2/csrc/array.h

```diff
@@ -161,30 +161,29 @@ class Array1 {
     return ans;
   }
 
-  // Resizes, copying old contents if we could not re-use the same memory
-  // location. It will always at least double the allocated size if it has to
-  // reallocate. See Region::num_bytes vs. Region::bytes_used.
-  // TODO(haowen): now we only support the case that the current array `curr`
-  // (i.e. the array that will be resized) covers the highest used index in
-  // the region, that is, for any array `a` uses this region,
-  // curr.byte_offset_ + curr.Dim() * curr.ElementSize() == region_->bytes_used
-  // >= a.byte_offset + a.Dim() * a.ElementSize()
+  /*
+   Extend size of array, copying old contents if we could not re-use the same memory
+   location. It will always at least double the allocated size if it has to
+   reallocate. See Region::num_bytes vs. Region::bytes_used.
+   We only support the case that the current array *this
+   (i.e. the array that will be resized) covers the highest used index in
+   the region, that is, for any array `a` uses this region,
+   curr.byte_offset_ + curr.Dim() * curr.ElementSize() == region_->bytes_used
+   >= a.byte_offset + a.Dim() * a.ElementSize()
+  */
   void Resize(int32_t new_size) {
-    K2_CHECK_EQ(byte_offset_ + Dim() * ElementSize(), region_->bytes_used);
-
-    if (new_size <= Dim()) return;
-
-    if (byte_offset_ + new_size * ElementSize() > region_->num_bytes) {
-      // always double the allocated size
-      auto tmp = NewRegion(Context(), 2 * region_->num_bytes);
-      // copy data
-      auto kind = GetMemoryCopyKind(*Context(), *Context());
-      MemoryCopy(tmp->data, region_->data, region_->bytes_used, kind);
-      // update meta info
-      dim_ = new_size;
-      tmp->bytes_used = byte_offset_ + new_size * ElementSize();
-      std::swap(tmp, region_);
+    if (new_size < dim_) {
+      K2_CHECK(new_size >= 0);
+    } else {
+      size_t cur_bytes_used = byte_offset_ + sizeof(T) * (size_t)dim_,
+          new_bytes_used = byte_offset_ + sizeof(T) * (size_t)new_size;
+      // the following avoids a situation where we overwrite data shared with
+      // other Array objects.  You can just do *this = Array1<T>(...) and
+      // overwrite *this with a new region if that's what you want.
+      K2_CHECK_EQ(cur_bytes_used, region_->bytes_used);
+      region_->Extend(new_bytes_used);
     }
+    dim_ = new_size;
```

Comment:
just to make sure, if `new_size < dim_`, here we will drop some elements in *this, right? But as `region_->bytes_used` is not changed, so the dropped memory before `region_->bytes_used` has a change that will not be used by any Array anymore (i.e., idle forever)?

Reply:
yes.  This is similar to `resize` in std::vector.  If needed we can later remove the constraint that the array must be
using the last byte.  The point is if the memory was used by other arrays, they may still be pointing to that memory.
This is basically to enable appending to arrays efficiently.